### PR TITLE
Optimize speed on Change Icon Dialog

### PR DESCRIPTION
**Pull request by zecarlos1957**
## Purpose

 dll/win32/shell32/dialogs/Dialogs.cpp
       Optimizing speed by eliminating DrawIconEx calls on hidden icons 

In line 1063 of listbox.c, we have an iterator from top_item to nb_items,
    which emits WM_DRAWITEM events to all list elements, even hidden ones.
    This upper bound could be clipped with LB_GETLISTBOXINFO.
    So we would only have WM_DRAWITEM events only for visible items,
    eliminating all surperful calls to hidden items.

### CODE REVIEW

File: dll/win32/shell32/dialogs/dialogs.cpp [hunk 1]
```diff
@@ -324,6 +324,12 @@ INT_PTR CALLBACK PickIconProc(
                 case ODA_SELECT:
                 case ODA_DRAWENTIRE:
                 {
+                    // Do not draw hiden icons
```
Comment (BieHDC) [inline dll/win32/shell32/dialogs/dialogs.cpp]: hidden*

Comment (zecarlos1957) [reply dll/win32/shell32/dialogs/dialogs.cpp]: Yes. Thank you

File: win32ss/user/ntuser/window.c [hunk 2]
```diff
@@ -3415,14 +3415,14 @@ NtUserGetListBoxInfo(
    }
 
    _SEH2_TRY
-   {
-      LB_DESCR *descr = ((PWND2LB)Wnd)->pLBiv;
+    {
```
Comment (tkreuzer) [inline win32ss/user/ntuser/window.c]: Please don't change indentation in the middle of the function.

File: dll/win32/shell32/dialogs/dialogs.cpp [hunk 3]
```diff
@@ -324,6 +324,12 @@ INT_PTR CALLBACK PickIconProc(
                 case ODA_SELECT:
                 case ODA_DRAWENTIRE:
                 {
+                    // Do not draw hidden icons
+                    UINT Visible = SendMessageW(pIconContext->hDlgCtrl, LB_GETLISTBOXINFO, 0, 0);
```
Comment (tkreuzer) [inline dll/win32/shell32/dialogs/dialogs.cpp]: Does this really optimize anything? Looks to me like additional processing for the most common case to "optimize" a corner case.

Comment (zecarlos1957) [reply dll/win32/shell32/dialogs/dialogs.cpp]: If we load the shell32.dll file, which contains 232 icons, we save (Item_Count - Top_Item - Visible_Items) calls to the DrawIconEx (...) function.
If the first item in the list is index 50, we eliminate 232-50-28 = 154 superfull calls to DrawIconEX (...).
There will certainly be more effective forms of optimization, but this is already noticeable in every refresh of the listbox.

**Comment by zecarlos1957** [comment]
In line 1063 of listbox.c, we have an iterator from top_item to nb_items,
    which emits WM_DRAWITEM events to all list elements, even hidden ones.
    This upper bound could be clipped with LB_GETLISTBOXINFO.
    So we would only have WM_DRAWITEM events only for visible items,
    eliminating all surfefulas calls to hidden items.
Is thar a better wai?

**Comment by binarymaster** [comment]
You have typos in commit title, here I fixed it:
```diff
-[USER32]Improvement to optimize speed on Change Icon Diaoog IDC_PICKICON_LIST
+[USER32] Improvement to optimize speed on Change Icon Dialog IDC_PICKICON_LIST
```

**Comment by zecarlos1957** [comment]
@binarymaster  Thanks

**Comment by tkreuzer** [comment]
About what performance improvement are we talking here? Do you have any measurements or any visible indicator, that it changes anything?

**Comment by KyleKatarn31** [comment]
And appart from the metrics, is this PR continued or abandonned ?

**Pull request closed without merging**